src/main/java/hudson/security/UserAttributesHelper.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.security;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.Attributes;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.GregorianCalendar;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Ease all the computations required to determine the user account optional attributes for creating
+ * the UserDetails that will be used by the SecurityRealm
+ */
+@Restricted(NoExternalUse.class)
+public class UserAttributesHelper {
+    // https://support.microsoft.com/en-us/help/305144/how-to-use-the-useraccountcontrol-flags-to-manipulate-user-account-pro
+    private static String ATTR_USER_ACCOUNT_CONTROL = "userAccountControl";
+    private static String ATTR_ACCOUNT_EXPIRES = "accountExpires";
+    // for Windows Server 2003-based domain
+    private static String ATTR_USER_ACCOUNT_CONTROL_COMPUTED = "msDS-User-Account-Control-Computed";
+    // for ADAM (Active Directory Application Mode), replace the ADS_UF_DISABLED
+    private static String ATTR_USER_ACCOUNT_DISABLED = "msDS-UserAccountDisabled";
+    // for ADAM, replace the ADS_UF_PASSWORD_EXPIRED
+    private static String ATTR_USER_PASSWORD_EXPIRED = "msDS-UserPasswordExpired";
+
+    // https://docs.microsoft.com/en-us/windows/desktop/adschema/a-accountexpires
+    // constant names follow the code in Iads.h
+    private static long ACCOUNT_NO_EXPIRATION = 0x7FFFFFFFFFFFFFFFL;
```

Comment:
```suggestion
    private static long ACCOUNT_NO_EXPIRATION = 0x7FFF_FFFF_FFFF_FFFFL;
```

This might help readability, though I'm not exactly sure if that syntax works with hex literals.

Reply:
That works ;-)